docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,545 @@
+
+<img title="" src="assets/04-state-transitions-restart-member.excalidraw.png" alt="" data-align="center">
+
+> NOTE: If the DB validation fails then data directory of the etcd-member is removed and etcd-member is removed from cluster membership, thus transitioning it to `New` state. The state transitions from `New` state are depicted by [this section](#addition-of-a-new-etcd-member-in-a-multi-node-etcd-cluster).
+
+### Deterministic Etcd Member Creation/Restart During Scale-Up
+
+**Bootstrap information:**
+
+When an etcd-member starts, then it needs to find out:
+
+* If it should join an existing cluster or start a new cluster.
+
+* If it should add itself as a `Learner` or directly start as a voting member.
+
+**Issue with the current approach**:
+
+*At present*, this is facilitated by three things:
+
+1. During scale-up, druid adds an annotation `gardener.cloud/scaled-to-multi-node` to the `StatefulSet`. Each etcd-members looks up this annotation.
+
+2. backup-sidecar attempts to fetch etcd cluster member-list and checks if this etcd-member is already part of the cluster.
+
+3. Size of the cluster by checking `initial-cluster` in the etcd config.
+
+Druid adds an annotation `gardener.cloud/scaled-to-multi-node` on the `StatefulSet` which is then shared by all etcd-members irrespective of the starting state of an etcd-member (as `Learner` or `Voting-Member`). This especially creates an issue for the current leader (often pod with index 0) during the scale-up of an etcd cluster as described in [this](https://github.com/gardener/etcd-backup-restore/issues/646) issue. 
+
+It has been agreed that the current solution to [this issue](https://github.com/gardener/etcd-backup-restore/issues/646) is a quick and dirty fix and needs to be revisited to be uniformly applied to all etcd-members. The authors propose to provide a more deterministic approach to scale-up using the `EtcdMember` resource.
+
+**New approach**
+
+Instead of adding an annotation `gardener.cloud/scaled-to-multi-node` on the `StatefulSet`, a new annotation `druid.gardener.cloud/create-as-learner` should be added by druid on an `EtcdMember` resource. This annotation will only be added to newly created members during scale-up.
+
+Each etcd-member should look at the following to deterministically compute the `bootstrap information` specified above:
+
+* `druid.gardener.cloud/create-as-learner` annotation on its respective `EtcdMember` resource. This new annotation will be honored in the following cases:
+  * When an etcd-member is created for the very first time.
+
+  * An etcd-member is restarted while it is in `Starting` state (`PendingLearner` and `Learner` sub-states).
+
+* Etcd-cluster member list.  to check if it is already part of the cluster. 
+
+* Existing etcd data directory and its validity.
+
+> **NOTE:** When the etcd-member gets promoted to a voting-member, then it should remove the annotation on its respective `EtcdMember` resource.
```

Comment:
what if it get restarted and failed to remove this annotation `druid.gardener.cloud/create-as-learner`

Reply:
Well, then when the member comes back up, if its data dir is valid, it doesn't need to see the annotation anymore, and will start as a follower as expected. But if its data dir is corrupt, then it will have to re-join the cluster as a learner. It's the same case today - if etcd-member-2 restarts just after it joined the cluster and its data dir gets corrupted, then it will rejoin the cluster as a learner, doesn't it?

Reply:
> if its data dir is valid, it doesn't need to see the annotation anymore

right now, we first check the scale-up condition first after data-dir validation is checked.
https://github.com/gardener/etcd-backup-restore/blob/64995e097fd75e9838887302cc41ff454557d511/pkg/initializer/initializer.go#L69-L91

Reply:
@ishan16696 don't worry about what is implemented currently. We have an opportunity to change the behavior and correct a few wrongs here, so I would implore you to look at it from that perspective.

If the data-sir is present and valid (revision matches the leader revision), do we really need to check the scale-up annotation? Scale-up annotation is required *only* when we need to determine whether the member needs to be added as a learner or not. If the data-dir is valid, then we have no need to add as a learner, hence no need to check the annotation. If the data-dir is invalid (corrupt), then we anyway need to clean up the data-dir and add as a learner so that it catches up with the leader, in the case of a multi-node cluster.

Reply:
> so I would implore you to look at it from that perspective.

yes, I already thought about this while working on scale-up fixes. I discussed this with Aaron sometime back that why we are first checking the scale-up then data-dir validation but right now I forgot what we discussed and the reason behind this.